File: Python/RHF.py
"""
    Copyright (C) 2015 Rocco Meli

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""

from matrices import *

import numpy.linalg as la

def RHF_step(basis,molecule,N,H,X,P_old,ee,verbose=False):
    """
    Restricted Hartree-Fock self-consistent field setp.

    INPUT:
        BASIS: basis set
        MOLECULE: molecule, collection of atom object
        N: Number of electrons
        H: core Hamiltonian
        X: tranformation matrix
        P_OLD: Old density matrix
        EE: List of electron-electron Integrals
        VERBOSE: verbose flag (set True to print everything on screen)
    """

    if verbose:
        print("\nDensity matrix P:")
        print(P_old)

    G = G_ee(basis,molecule,P_old,ee) # Compute electron-electron interaction matrix

    if verbose:
        print("\nG matrix:")
        print(G)

    F = H + G # Compute Fock matrix

    if verbose:
        print("\nFock matrix:")
        print(F)

    Fx = np.dot(X.conj().T,np.dot(F,X)) # Compute Fock matrix in the orthonormal basis set (S=I in this set)

    if verbose:
        print("\nFock matrix in orthogonal orbital basis:")
        print(Fx)

    e, Cx = la.eigh(Fx) # Compute eigenvalues and eigenvectors of the Fock matrix

    # Sort eigenvalues from smallest to highest (needed to compute P correctly)
    idx = e.argsort()
    e = e[idx]
    Cx = Cx[:,idx]

    if verbose:
        print("\nCoefficients in orthogonal orbital basis:")
        print(Cx)

    e = np.diag(e) # Extract orbital energies as vector

    if verbose:
        print("\nEnergies in orthogonal orbital basis:")
        print(e)

    C = np.dot(X,Cx) # Transform coefficient matrix in the orthonormal basis to the original basis

    if verbose:
        print("\nCoefficients:")
        print(C)

    Pnew = P_density(C,N) # Compute the new density matrix

    return Pnew, F, e

def delta_P(P_old,P_new):
    """
    Compute the difference between two density matrices.

    INTPUT:
        P_OLD: Olde density matrix
        P_NEW: New density matrix
    OUTPUT:
        DELTA: difference between the two density matrices

    Source:
        Modern Quantum Chemistry
        Szabo and Ostlund
        Dover
        1989
    """
    delta = 0

    n = P_old.shape[0]

    for i in range(n):
        for j in range(n):
            delta += (P_old[i,j] - P_new[i,j])**2

    return (delta / 4.)**(0.5)

def energy_el(P,F,H):
    """
    Compute electronic energy.

    INPUT:
        P: Density matrix
        F: Fock matrix
        H: Core Hamiltonian

    Source:
        Modern Quantum Chemistry
        Szabo and Ostlund
        Dover
        1989
    """

    # Size of the basis set
    K = P.shape[0]

    E = 0

    for i in range(K):
        for j in range(K):
            E += 0.5 * P[i,j] * (H[i,j] + F[i,j])

    return E

def energy_n(molecule):
    """
    Compute nuclear energy (classical nucleus-nucleus repulsion)

    INPUT:
        MOLECULE: molecule, as a collection of atoms
    OUTPUT:
        ENERGY_N: Nuclear energy
    """

    en = 0

    for i in range(len(molecule)):
        for j in range(i+1,len(molecule)):
            # Select atoms from molecule
            atomi = molecule[i]
            atomj = molecule[j]

            # Extract distance from atom
            Ri = np.asarray(atomi.R)
            Rj = np.asarray(atomj.R)

            en += atomi.Z * atomj.Z / la.norm(Ri - Rj)

    return en

def energy_tot(P,F,H,molecule):
    """
    Compute total energy (electronic plus nuclear).

    INPUT:
        P: Density matrix
        F: Fock matrix
        H: Core Hamiltonian
        MOLECULE: molecule, as a collection of atoms
    OUTPUT:
        ENERGY_TOT: total energy
    """
    return energy_el(P,F,H) + energy_n(molecule)

File: Fortran90/RHF.f90
! --------------------------------------------------------------------
!
! Copyright (C) 2015 Rocco Meli
!
! This program is free software: you can redistribute it and/or modify
! it under the terms of the GNU General Public License as published by
! the Free Software Foundation, either version 3 of the License, or
! (at your option) any later version.
!
! This program is distributed in the hope that it will be useful,
! but WITHOUT ANY WARRANTY; without even the implied warranty of
! MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
! GNU General Public License for more details.
!
! You should have received a copy of the GNU General Public License
! along with this program.  If not, see <http://www.gnu.org/licenses/>.
!
! ---------------------------------------------------------------------

MODULE RHF
    ! ----------------------------------------------
    ! RESTRICTED HARTREE-FOCK
    ! ----------------------------------------------
    !
    ! Self-consistent solution of Roothaan equations
    ! (Restricted Hartree-Fock)
    !
    ! ----------------------------------------------

    USE UTILS, only: EIGS, print_real_matrix, print_ee_list
    USE DENSITY, only: P_density, delta_P
    USE ELECTRONIC, only: G_ee, EE_list
    USE ENERGY, only: E_tot
    USE CONSTANTS
    USE OVERLAP, only: S_overlap, X_transform
    USE CORE, only: H_core

    IMPLICIT NONE

    CONTAINS

        ! --------
        ! SCF STEP
        ! --------
        SUBROUTINE RHF_step(Kf,Ne,H,X,ee,Pold,Pnew,F,orbitalE,verbose)
            ! ----------------------------------------------------------------------
            ! Perform a single step of the SCF procedure to solve Roothan equations.
            ! ----------------------------------------------------------------------
            !
            ! Source:

            IMPLICIT NONE

            ! INPUT
            INTEGER, intent(in) :: Kf                           ! Number of basis functions
            !INTEGER, intent(in) :: Nn                           ! Number of nuclei
            INTEGER, intent(in) :: Ne                           ! Number of electrons
            !REAL*8, dimension(Kf,3), intent(in) :: basis_D      ! Contraction coefficients
            !REAL*8, dimension(Kf,3), intent(in) :: basis_A      ! Contraction exponential coefficients
            !INTEGER, dimension(Kf,3), intent(in) :: basis_L     ! Basis function angular momenta
            !REAL*8, dimension(Kf,3), intent(in) :: basis_R      ! Basis set centers
            !REAL*8, dimension(Nn,3), intent(in) :: Rn           ! Nuclear coordinates
            !INTEGER, dimension(Nn), intent(in) :: Zn            ! Nuclear charges
            REAL*8, dimension(Kf,Kf), intent(in) :: H           ! Core Hamiltonian
            REAL*8, dimension(Kf,Kf), intent(in) :: X           ! Transformation maxrix
            REAL*8, dimension(Kf,Kf), intent(in) :: Pold        ! Old density matrix
            REAL*8, dimension(Kf,Kf,Kf,Kf), intent(in) :: ee    ! Electron-electron list
            LOGICAL, intent(in) :: verbose                      ! Flag to print matrices

            ! INTERMEDIATE VARIABLES
            REAL*8, dimension(Kf,Kf) :: G                       ! Electron-electron repulsion matrix
            REAL*8, dimension(Kf,Kf) :: Fx                      ! Fock matrix in the orthogonal basis set
            REAL*8, dimension(Kf,Kf) :: Cx                      ! Coefficient matrix in the orthogonal basis set
            REAL*8, dimension(Kf,Kf) :: C                       ! Coefficient matrix in the original basis set

            ! INPUT / OUTPUT
            REAL*8, dimension(Kf,Kf), intent(out) :: Pnew       ! New density matrix
            REAL*8, dimension(Kf,Kf), intent(out) :: F          ! Fock operator
            REAL*8, dimension(Kf), intent(out) :: orbitalE      ! Orbital energies

            IF (verbose) THEN
                WRITE(*,*)
                WRITE(*,*) "Density matrix P:"
                CALL print_real_matrix(Kf,Kf,Pold)
            END IF

            CALL G_ee(Kf,ee,Pold,G) ! Compute new electron-electron repulsion matrix G

            IF (verbose) THEN
                WRITE(*,*)
                WRITE(*,*) "Electron-electron repulsion matrix G:"
                CALL print_real_matrix(Kf,Kf,G)
            END IF

            F = H + G ! Compute new Fock operator

            IF (verbose) THEN
                WRITE(*,*)
                WRITE(*,*) "Fock matrix F:"
                CALL print_real_matrix(Kf,Kf,F)
            END IF

            Fx = MATMUL(TRANSPOSE(X),MATMUL(F,X))

            IF (verbose) THEN
                WRITE(*,*)
                WRITE(*,*) "Fock matrix in orthogonal orbital basis Fx:"
                CALL print_real_matrix(Kf,Kf,Fx)
            END IF

            ! ------------------------------------
            ! Solve orthogonalized Roothan eqution
            ! ------------------------------------
            CALL EIGS(Kf,Fx,Cx,orbitalE) ! Compute coefficients (in orthonormal basis) and orbital energies

            IF (verbose) THEN
                WRITE(*,*)
                WRITE(*,*) "Coefficients in orthogonal orbital basis Cx:"
                CALL print_real_matrix(Kf,Kf,Cx)
            END IF

            IF (verbose) THEN
                WRITE(*,*)
                WRITE(*,*) "Orbital energies:"
                CALL print_real_matrix(Kf,1,orbitalE)
            END IF

            C = MATMUL(X,Cx) ! Compute coefficients in the original basis

            IF (verbose) THEN
                WRITE(*,*)
                WRITE(*,*) "Coefficients:"
                CALL print_real_matrix(Kf,Kf,C)
            END IF

            CALL P_density(Kf,Ne,C,Pnew) ! Compute new density

        END SUBROUTINE RHF_step


        ! ---------
        ! SCF Cycle
        ! ---------
        SUBROUTINE RHF_SCF(Kf,c,Ne,Nn,basis_D,basis_A,basis_L,basis_R,Zn,Rn,final_E,verbose)
            ! --------------------------------
            ! Compute total energy (SCF cycle)
            ! --------------------------------

            IMPLICIT NONE

            ! INPUT
            INTEGER, intent(in) :: Kf                                   ! Basis set size
            INTEGER, intent(in) :: c                                    ! Number of contractions
            INTEGER, intent(in) :: Ne                                   ! Number of electrons
            INTEGER, intent(in) :: Nn                                   ! Number of nuclei
            INTEGER, dimension(Kf,3), intent(in) :: basis_L             ! Angular momenta of basis set Gaussians
            REAL*8, dimension(Kf,3), intent(in) :: basis_R              ! Centers of basis set Gaussians
            REAL*8, dimension(Kf,c), intent(in) :: basis_D, basis_A     ! Basis set coefficients
            REAL*8, dimension(Nn,3), intent(in) :: Rn                   ! Nuclear positions
            INTEGER, dimension(Nn), intent(in) :: Zn                    ! Nuclear charges
            LOGICAL, intent(in) :: verbose                              ! Verbose flag

            ! OUTPUT
            REAL*8,intent(out) :: final_E ! Converged total energy

            ! --------
            ! Matrices
            ! --------

            REAL*8, dimension(Kf,Kf) :: S     ! Overlap matrix
            REAL*8, dimension(Kf,Kf) :: X     ! Transformation matrix

            REAL*8, dimension(Kf,Kf) :: Hc    ! Core Hamiltonian

            REAL*8, dimension(Kf,Kf) :: Pold  ! Old density matrix
            REAL*8, dimension(Kf,Kf) :: Pnew  ! New density matrix

            REAl*8, dimension(Kf,Kf,Kf,Kf) :: ee ! List of electron-electron integrals

            REAL*8, dimension(Kf,Kf) :: F     ! Fock matrix
            REAL*8, dimension(Kf) :: E           ! Orbital energies

            ! --------------
            ! SCF parameters
            ! --------------

            LOGICAL :: converged                    ! Convergence parameter
            INTEGER, PARAMETER :: maxiter = 250     ! Maximal number of iterations (TODO: user defined maxiter)
            INTEGER :: step                         ! SCF steps counter

            !!!
            !!! NEVER INITIALIZE ON DECLARATION
            !!! "A local variable that is initialized when declared has an implicit save attribute.
            !!! The variable is initialized only the first time the unction is called.
            !!! On subsequent calls the old value is retained."
            !!!

            converged = .FALSE.
            step = 0

            ! -----------------
            ! HF initialization
            ! -----------------

            CALL S_overlap(Kf,c,basis_D,basis_A,basis_L,basis_R,S) ! Compute overlap matrix

            IF (verbose) THEN
                WRITE(*,*) "Overlap matrix S:"
                CALL print_real_matrix(Kf,Kf,S)
            END IF

            CALL X_transform(Kf,S,X) ! Compute transformation matrix

            IF (verbose) THEN
                WRITE(*,*) "Transformation matrix X:"
                CALL print_real_matrix(Kf,Kf,X)
            END IF

            CALL H_core(Kf,c,Nn,basis_D,basis_A,basis_L,basis_R,Rn,Zn,Hc,verbose)

            IF (verbose) THEN
                WRITE(*,*) "Core Hamiltonian Hc:"
                CALL print_real_matrix(Kf,Kf,Hc)
            END IF

            CALL EE_list(Kf,c,basis_D,basis_A,basis_L,basis_R,ee)

            IF (verbose) THEN
                WRITE(*,*) "Two-electron integrals:"
                CALL print_ee_list(Kf,ee)
            END IF

            Pold(:,:) = 0.0D0
            Pnew(:,:) = 0.0D0

            ! ---------
            ! SCF cycle
            ! ---------

            DO WHILE ((converged .EQV. .FALSE.) .AND. step .LT. maxiter)
                step = step + 1

                IF (verbose) THEN
                    WRITE(*,*)
                    WRITE(*,*)
                    WRITE(*,*)
                    WRITE(*,*) "--------"
                    WRITE(*,*) "SCF step #", step
                    WRITE(*,*) "--------"
                END IF

                CALL RHF_step(Kf,Ne,Hc,X,ee,Pold,Pnew,F,E,verbose)

                IF (verbose) THEN
                    WRITE(*,*   )
                    WRITE(*,*) "Total energy:", E_tot(Kf,Nn,Rn,Zn,Pold,F,Hc)
                END IF

                IF ( delta_P(Kf,Pold,Pnew) < 1.0e-12) THEN
                    converged = .TRUE.

                    final_E = E_tot(Kf,Nn,Rn,Zn,Pold,F,Hc)

                    IF (verbose) THEN
                        WRITE(*,*)
                        WRITE(*,*) "SCF cycle converged!"
                        WRITE(*,*)
                        WRITE(*,*)
                        WRITE(*,*)
                        WRITE(*,*) "TOTAL ENERGY:", final_E
                    END IF
                END IF

                Pold = Pnew

            END DO ! SCF

            IF (converged .EQV. .FALSE.) THEN
                WRITE(*,*)
                WRITE(*,*) "SCF NOT CONVERGED!"
                CALL EXIT(-1)
            END IF

        END SUBROUTINE RHF_SCF


END MODULE RHF
